consensus/fsm_handler.go

```diff
@@ -0,0 +1,102 @@
+		return m.HandleSync(msg)
+
+	case string(coreTypes.StateMachineState_Consensus_Synced):
+		return m.HandleSynced(msg)
+
+	case string(coreTypes.StateMachineState_Consensus_Pacemaker):
+		return m.HandlePacemaker(msg)
+	}
+
+	return nil
+}
+
+// Bootrstapped mode is when the node (validator or non-valdiator) is out of sync with the rest of the network
+// This mode is a transition mode from node bootstrappin to node being out-of-sync.
+func (m *consensusModule) HandleBootstrapped(msg *messaging.StateMachineTransitionEvent) error {
+	m.logger.Debug().Msg("FSM is in bootstrapped state, so it is out of sync, and transitions to unsynched mode")
+	if err := m.GetBus().GetStateMachineModule().SendEvent(coreTypes.StateMachineEvent_Consensus_IsUnsynched); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Unsynched mode is when the node (validator or non-valdiator) is out of sync with the rest of the network
+// This mode is a transition mode from node being up-to-date (i.e. Pacemaker mode, Synched mode) to the latest state to node being out-of-sync
+// As soon as node transitions to this mode, it will transition to the sync mode.
+func (m *consensusModule) HandleUnsynched(msg *messaging.StateMachineTransitionEvent) error {
+	m.logger.Debug().Msg("FSM is in Unsyched state, as node is out of sync sending syncmode event to start syncing")
+	if err := m.GetBus().GetStateMachineModule().SendEvent(coreTypes.StateMachineEvent_Consensus_IsSyncing); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Sync mode is when the node (validator or non-valdiator) is syncing with the rest of the network
+func (m *consensusModule) HandleSync(msg *messaging.StateMachineTransitionEvent) error {
+	m.logger.Debug().Msg("FSM is in Sync Mode, starting syncing")
+
+	m.stateSync.AggregateMetadataResponses()
+
+	// try sycing until node is synced
+	// CONSIDER: consider putting a limit on number of tries, or timeout
+	err := m.stateSync.Sync()
+	for err != nil {
+		err = m.stateSync.Sync()
+	}
+
+	if m.IsValidator() {
+		m.logger.Debug().Msg("Validator node synced to the latest state with the rest of the peers")
+		if err := m.GetBus().GetStateMachineModule().SendEvent(coreTypes.StateMachineEvent_Consensus_IsCaughtUpValidator); err != nil {
+			return err
+		}
+	} else {
+		m.logger.Debug().Msg("Non-Validator synced to the latest state with the rest of the peers")
+		if err := m.GetBus().GetStateMachineModule().SendEvent(coreTypes.StateMachineEvent_Consensus_IsCaughtUpNonValidator); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Currently, FSM never transition to this state and a non-validator node always stays in syncmode.
+// CONSIDER: when a non-validator sync is implemented, maybe there is a case that requires transitioning to this state
+func (m *consensusModule) HandleSynced(msg *messaging.StateMachineTransitionEvent) error {
+	m.logger.Debug().Msg("FSM of non-validator node is in Synced mode")
```

Comment:
Can you add a TODO to check that this never happens if you're `IsValidator` is true?

Reply:
I think this is not correct. Because, a validator transitions to this state after it synchs with the rest of the network. Validator stays in Pacemaker mode until it receives a block proposal that is higher than its current height, and it transitions to the unsynched state. 

Lmk what do you think.

Reply:
I think either works. Seems like you designed it so that a validator goes from synched -> pacemaker, right?

If so, feel free to update the comment appropriately.

Reply:
<img width="1040" alt="Screenshot 2023-03-10 at 11 33 56 AM" src="https://user-images.githubusercontent.com/1892194/224411154-63af5b62-e247-46b2-8753-f4e569c87aaf.png">